test/RequestChannelTest.cpp

```diff
@@ -88,11 +88,299 @@ TEST(RequestChannelTest, RequestOnDisconnectedClient) {
+        std::stringstream ss;
+        ss << "Requester stream: " << v << " of " << requesterRangeEnd;
+        std::string s = ss.str();
+        return Payload(s, "metadata");
+      });
+
+  requester->requestChannel(requesterFlowable)
+      ->map([](auto p) { return p.moveDataToString(); })
+      ->subscribe(requestSubscriber);
+
+  // finish streaming from Responder
+  requestSubscriber->awaitTerminalEvent();
+  requestSubscriber->assertSuccess();
+  requestSubscriber->assertValueCount(10);
+  requestSubscriber->assertValueAt(0, "Responder stream: 1 of 10");
+  requestSubscriber->assertValueAt(9, "Responder stream: 10 of 10");
+
+  // Responder stream is closed, Requester continues
+  responderSubscriber->request(50);
+  responderSubscriber->awaitTerminalEvent();
+  responderSubscriber->assertSuccess();
+  responderSubscriber->assertValueCount(100);
+  responderSubscriber->assertValueAt(0, "Requester stream: 1 of 100");
+  responderSubscriber->assertValueAt(99, "Requester stream: 100 of 100");
+}
+
+TEST(RequestChannelTest, FlowControl) {
+  int64_t responderRange = 10;
+  int64_t responderSubscriberInitialRequest = 0;
+
+  auto responder = std::make_shared<TestChannelResponder>(
+      responderRange, responderSubscriberInitialRequest);
+
+  folly::ScopedEventBaseThread worker;
+  auto server = makeServer(responder);
+  auto client = makeClient(worker.getEventBase(), *server->listeningPort());
+  auto requester = client->getRequester();
+
+  auto requestSubscriber = TestSubscriber<std::string>::create(1);
+  auto responderSubscriber = responder->getChannelSubscriber();
+
+  int64_t requesterRangeEnd = 10;
+
+  auto requesterFlowable =
+      Flowables::range(1, requesterRangeEnd)->map([&](int64_t v) {
+        std::stringstream ss;
+        ss << "Requester stream: " << v << " of " << requesterRangeEnd;
+        std::string s = ss.str();
+        return Payload(s, "metadata");
+      });
+
+  requester->requestChannel(requesterFlowable)
+      ->map([](auto p) { return p.moveDataToString(); })
+      ->subscribe(requestSubscriber);
+
+  responderSubscriber->awaitValueCount(1);
+  requestSubscriber->awaitValueCount(1);
+
+  for (int i = 2; i <= 10; i++) {
+    requestSubscriber->request(1);
+    responderSubscriber->request(1);
+
+    responderSubscriber->awaitValueCount(i);
+    requestSubscriber->awaitValueCount(i);
+
+    requestSubscriber->assertValueCount(i);
+    responderSubscriber->assertValueCount(i);
+  }
+
+  requestSubscriber->awaitTerminalEvent();
+  responderSubscriber->awaitTerminalEvent();
+
+  requestSubscriber->assertSuccess();
+  responderSubscriber->assertSuccess();
+
+  requestSubscriber->assertValueAt(0, "Responder stream: 1 of 10");
+  requestSubscriber->assertValueAt(9, "Responder stream: 10 of 10");
+
+  responderSubscriber->assertValueAt(0, "Requester stream: 1 of 10");
+  responderSubscriber->assertValueAt(9, "Requester stream: 10 of 10");
+}
+
+TEST(RequestChannelTest, DISABLED_CancelFromRequester) {
```

Comment:
Yup, the leak is definitely a result of running the cancel test.

Reply:
Leak when CancelFromRequester runs:
https://travis-ci.org/rsocket/rsocket-cpp/jobs/277537750#L1984

I can't reproduce this locally, does anyone have any ideas?

Reply:
I would guess the leak is that the allocated string is put in the TestSubscriber which is held by the TestChannelResponder which is held by the rsocket server? Can you verify that the responder is released?
Also you can locally compile the yarpl library with Dylan's new switch which will allow you to call "print_ref_counted_objects" at the end of the test so you can see what refcounted objects are still alive.